### PR TITLE
ceph: support specifying only public network or only cluster network or both

### LINKED ISSUE #7436: Ceph+Multus: Support specifying only cluster selector (defer to primary CNI for public)
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:** Allow specifying a Multus NAD for only cluster selector, defer to default CNI plugin for public network.

Cluster operators should not need a Multus NAD for the public network. 

Please advise if I am missing an important detail :-)

**What is use case behind this feature:** It seems the most common use of Multus in Rook is to add a cluster network for Ceph, and the public network should be the default pod network which is already managed by the primary CNI plugin (e.g. Calico, etc.).

Currently, Rook supports either both, or only `public` to be specified under `CephCluster.spec.network.selectors`. When both are specified, they both need to be Multus NADs. It is not intuitive (I have not been able to validate that it is even possible) to create a Multus NAD that effectively defers to the primary CNI. (note: It would be satisfactory that, if it is supported that Multus can defer to primary CNI, this issue be closed with reference to docs on how to do this).

Specifically, It gives me anxiety to need to specify a Multus NAD that configures an attachment into my default pod network when I have already granted Calico control of the pod network. Do I specify the same range? Do I need a distinct range in the same subnet? Do I need to encode every single network device from each node?  etc. etc. 

**Environment**:
Rook version: 1.5.8
Multus: `ghcr.io/k8snetworkplumbingwg/multus-cni:stable`
CNI: Calico (node version v3.17.1)

### PR DESCRIPTION
**Pull request by rohan47**
**Description of your changes:**
With this change, the Rook ceph cluster can have only a public network or
only a cluster network or both can be specified. Prior to this commit if
only a public network was specified, it was copied to the cluster network.

Also, cluster network will only be applied to Osd pods.

**Which issue is resolved by this Pull Request:**
Resolves #7436

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by rohan47** [comment]
The code is not tested on an actual cluster so added the do-no-merge label

File: pkg/operator/k8sutil/network.go [hunk 1]
```diff
@@ -77,13 +79,13 @@ func ApplyMultus(net rookv1.NetworkSpec, objectMeta *metav1.ObjectMeta) error {
 			shortSyntax = true
 		}
 
-		isCsi := strings.Contains(objectMeta.Labels["app"], "csi-")
```
Comment (travisn) [inline pkg/operator/k8sutil/network.go:80]: Why the change to the csi pods? They no longer need the cluster network? Whey did we have the filter before?

File: pkg/operator/k8sutil/network.go [hunk 2]
```diff
@@ -77,13 +79,13 @@ func ApplyMultus(net rookv1.NetworkSpec, objectMeta *metav1.ObjectMeta) error {
...
-		if isCsi {
+		isOSD := strings.Contains(objectMeta.Labels["app"], "osd")
```
Comment (travisn) [inline pkg/operator/k8sutil/network.go:82]: Let's not hard-code the apps that need to be excluded. How about defining a slide of the apps to exclude, then this implementation can iterate over the slice. Otherwise, we have to add every exception as a new line of code instead of just adding to the slice when there is a new exception.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/k8sutil/network.go (hunk 1 above)
Comment (rohan47) [reply pkg/operator/k8sutil/network.go:80]: We had the filter to apply only the public network to the csi pods

Comment (travisn) [reply pkg/operator/k8sutil/network.go:80]: So now OSDs are the only ones that apply the cluster network, right? This means mon, mgr, and other daemons were applying both before, but will only apply the public network now.

Comment (rohan47) [reply pkg/operator/k8sutil/network.go:80]: Yes, Correct

**Comment by travisn** [comment]
@rohan47 The only thing left on this is to test?
@ctrahey Want to help verify it?

**Comment by rohan47** [comment]
> @rohan47 The only thing left on this is to test?
Yes, I am not able to get a working bare metal cluster.
> @ctrahey Want to help verify it?

**Comment by ctrahey** [comment]
I can test, yes.
Might take me a couple days but I'll report back - thanks for the effort here!

**Comment by ctrahey** [comment]
I encountered build errors. Output [here](https://gist.github.com/ctrahey/98f8b5ed2cadbb37f4a15a010c63a241).

If you already have an image somewhere I can try with that - I can run the operator x86 or ARM.

However - I should let you know that we're moving and sometime tomorrow I will power down the cluster, de-rack everything, and truck it, and I don't yet have a specific date when I know I can make time to put it back together. I hope by mid-May.

File: pkg/operator/ceph/config/network_test.go [hunk 3]
```diff
@@ -50,19 +50,14 @@ func TestGenerateNetworkSettings(t *testing.T) {
 	assert.Error(t, err)
 
 	//
-	// TEST 2: single dedicated networks
+	// TEST 2: single dedicated network
```
Comment (leseb) [inline pkg/operator/ceph/config/network_test.go:53]: We actually want to test `cluster_network` alone too.

**Review by leseb — CHANGES_REQUESTED** [review]

**Comment by rohan47** [comment]
I have tested the code on a baremetal cluster having 2 nics on each node.
Tested two scenarios:
1. Using public and cluster network
2. Using only the cluster network

I confirmed the connections using `ss` :
e.g., connections of an OSD using cluster network only
```
$ oc rsh rook-ceph-osd-0-66cd885dfd-stpnw
sh-4.4# ss -tlnpa
State                 Recv-Q                Send-Q                               Local Address:Port                                 Peer Address:Port
ESTAB                 0                     0                                    192.168.232.4:6802                                192.168.232.3:43028                users:(("ceph-osd",pid=1,fd=70))
ESTAB                 0                     0                                     10.129.9.168:59538                                10.128.0.184:6804                 users:(("ceph-osd",pid=1,fd=62))
ESTAB                 0                     0                                     10.129.9.168:6800                                 10.129.9.161:59336                users:(("ceph-osd",pid=1,fd=74))
ESTAB                 0                     0                                     10.129.9.168:43860                               172.30.134.63:3300                 users:(("ceph-osd",pid=1,fd=56))
ESTAB                 0                     0                                     10.129.9.168:6800                                 10.129.9.171:52796                users:(("ceph-osd",pid=1,fd=72))
ESTAB                 0                     0                                    192.168.232.4:52060                               192.168.232.2:6802                 users:(("ceph-osd",pid=1,fd=65))
ESTAB                 0                     0                                     10.129.9.168:6804                                  10.130.1.18:43602                users:(("ceph-osd",pid=1,fd=68))
ESTAB                 0                     0                                    192.168.232.4:6806                                192.168.232.2:56632                users:(("ceph-osd",pid=1,fd=64))
ESTAB                 0                     0                                     10.129.9.168:6800                                  10.130.1.12:34782                users:(("ceph-osd",pid=1,fd=73))
ESTAB                 0                     0                                     10.129.9.168:52446                                10.128.0.180:6800                 users:(("ceph-osd",pid=1,fd=66))
ESTAB                 0                     0                                     10.129.9.168:6804                                 10.128.0.184:56234                users:(("ceph-osd",pid=1,fd=69))
ESTAB                 0                     0                                     10.129.9.168:6800                                  10.130.1.19:55504                users:(("ceph-osd",pid=1,fd=71))
ESTAB                 0                     0                                    192.168.232.4:6806                                192.168.232.3:43222                users:(("ceph-osd",pid=1,fd=67))
ESTAB                 0                     0                                    192.168.232.4:44222                               192.168.232.3:6806                 users:(("ceph-osd",pid=1,fd=60))
ESTAB                 0                     0                                    192.168.232.4:42920                               192.168.232.2:6806                 users:(("ceph-osd",pid=1,fd=61))
ESTAB                 0                     0                                     10.129.9.168:59264                                 10.130.1.18:6804                 users:(("ceph-osd",pid=1,fd=63))
```

verified that csi, cephfs and rbd are functional with these changes

File: pkg/operator/ceph/config/network_test.go (hunk 3 above)
Comment (rohan47) [reply pkg/operator/ceph/config/network_test.go]: Done

File: pkg/operator/k8sutil/network_test.go [hunk 4]
```diff
@@ -29,18 +29,33 @@ func TestNetwork_ApplyMultusShort(t *testing.T) {
...
+	// TEST 2: Osd pod
+	//
+	objMetaOsd := metav1.ObjectMeta{}
+	objMetaOsd.Labels = map[string]string{
+		"app": "rook-ceph-osd",
+	}
+	err = ApplyMultus(net, &objMetaOsd)
+	assert.NoError(t, err)
+	assert.Contains(t, objMetaOsd.Annotations["k8s.v1.cni.cncf.io/networks"], "macvlan@net1")
+	assert.Contains(t, objMetaOsd.Annotations["k8s.v1.cni.cncf.io/networks"], "macvlan@net2")
```
[the remark this answers is not in the document]
Comment (rohan47) [reply pkg/operator/k8sutil/network_test.go]: @BlaineEXE I have updated the tests

File: pkg/operator/ceph/config/network.go [hunk 5]
```diff
@@ -50,10 +50,8 @@ func generateNetworkSettings(clusterdContext *clusterd.Context, namespace string
 	cephNetworks := []Option{}
 
 	for _, selectorKey := range NetworkSelectors {
-		// This means only "public" was specified and thus we use the same subnet for cluster too
+		// This means only one network selector was specified
 		if _, ok := networkSelectors[selectorKey]; !ok {
-			cephNetworks = append(cephNetworks, cephNetworks[0])
-			cephNetworks[1].Option = fmt.Sprintf("%s_network", selectorKey)
 			continue
 		}
```
Comment (BlaineEXE) [inline pkg/operator/ceph/config/network.go:56]: It's not clear to me how this means only one network selector is specified? What is "This" referring to, and how do we know only one is specified in this loop?

File: pkg/operator/ceph/config/network_test.go [hunk 6]
```diff
@@ -31,45 +31,92 @@ import (
...
+					_, err := ctx.NetworkClient.NetworkAttachmentDefinitions(ns).Create(ctxt, &test.nw[i], metav1.CreateOptions{})
+					assert.NoError(t, err)
+				}
+				cephNetwork, err := generateNetworkSettings(ctx, ns, test.netSelector)
+				assert.NoError(t, err)
+				assert.ElementsMatch(t, cephNetwork, test.expectedNw, fmt.Sprintf("networks: %+v", cephNetwork))
+			} else {
+				_, err := generateNetworkSettings(ctx, ns, test.netSelector)
+				assert.Error(t, err)
+			}
```
Comment (BlaineEXE) [inline pkg/operator/ceph/config/network_test.go:111]: This is far too complex for a unit test. When I suggested to use `t.Run` I did not mean to run templated tests in a loop. Looped tests are fine for simple functions with small inputs and outputs, but not complex cases where we need if-else to check outputs. 

Just run the tests with `t.Run` directly.

```
func TestMyFunc(t *testing.T) {
    t.Run("subtest 1", func(t *testing.T) {
       // subtest stuff
    })
    
    t.Run("subtest 2", func(t *testing.T) {
        // other subtest stuff
    })
}

File: pkg/operator/k8sutil/network_test.go [hunk 7]
```diff
@@ -29,18 +29,45 @@ func TestNetwork_ApplyMultusShort(t *testing.T) {
...
+			err := ApplyMultus(net, &objMeta)
+			assert.NoError(t, err)
+
+			assert.Contains(t, objMeta.Annotations, "k8s.v1.cni.cncf.io/networks")
+			for _, want := range test.want {
+				assert.Contains(t, objMeta.Annotations["k8s.v1.cni.cncf.io/networks"], want)
+			}
+			for _, notWant := range test.notWant {
+				assert.NotContains(t, objMeta.Annotations["k8s.v1.cni.cncf.io/networks"], notWant)
+			}
```
Comment (BlaineEXE) [inline pkg/operator/k8sutil/network_test.go:67]: It's better to verify the output is **exactly** what is desired. This code as written does not protect future contributions from adding undesirable content to the annotations.

This is a more exact test.
```go
assert.ElementsMatch(t, objectMeta.Annotations["k8s.v1.cni.cncf.io/networks"], want)
```

File: pkg/operator/k8sutil/network_test.go [hunk 8]
```diff
@@ -53,6 +80,9 @@ func TestNetwork_ApplyMultusJSON(t *testing.T) {
 	}
 
 	objMeta := metav1.ObjectMeta{}
+	objMeta.Labels = map[string]string{
+		"app": "rook-ceph-osd",
+	}
 	err := ApplyMultus(net, &objMeta)
 	assert.NoError(t, err)
```
Comment (BlaineEXE) [inline pkg/operator/k8sutil/network_test.go:88]: I realize the  test below is pre-existing, but my previous comment still applies. It is better to test that the annotations are exactly as desired and not merely that they "contain" the things desired because the function could be modified in the future to add undesirable content. It would be good to modify the below to check that the annotation content is exactly as desired or at minimum verify that there are only 2 items present under `objMeta.Annotations["k8s.v1.cni.cncf.io/networks"]`

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: pkg/operator/k8sutil/network_test.go (hunk 7 above)
Comment (rohan47) [reply pkg/operator/k8sutil/network_test.go:67]: we are comparing strings i.e., `objectMeta.Annotations["k8s.v1.cni.cncf.io/networks"]` and `want`
can we use the following
```go
assert.Equal(t, objectMeta.Annotations["k8s.v1.cni.cncf.io/networks"], want)
```

Comment (rohan47) [reply pkg/operator/k8sutil/network_test.go:67]: or
```go
assert.Exactly(t, objMeta.Annotations, want)
```

Comment (rohan47) [reply pkg/operator/k8sutil/network_test.go]: done

File: pkg/operator/k8sutil/network.go [hunk 9]
```diff
@@ -104,6 +109,11 @@ func ApplyMultus(net rookv1.NetworkSpec, objectMeta *metav1.ObjectMeta) error {
 	return nil
 }
 
+// getClusterNwApps returns
+func getClusterNwApps() []string {
```
[the remark this answers is not in the document]
Comment (rohan47) [reply pkg/operator/k8sutil/network.go]: done

File: pkg/operator/ceph/config/network.go [hunk 10]
```diff
@@ -50,10 +50,8 @@ func generateNetworkSettings(clusterdContext *clusterd.Context, namespace string
 	cephNetworks := []Option{}
 
 	for _, selectorKey := range NetworkSelectors {
-		// This means only "public" was specified and thus we use the same subnet for cluster too
+		// When only one network selector is specified
```
Comment (BlaineEXE) [inline pkg/operator/ceph/config/network.go:53]: I still don't really understand this comment. I think it's safe to remove, tbh. All this code does is skip processing selectors if they aren't specified.

**Review by BlaineEXE — APPROVED** [review]
Looks good. I mentioned a comment that still doesn't make sense to me, but I think we can merge this and consider the comment separately since it's a pretty small nit.

File: pkg/operator/ceph/config/network.go (hunk 10 above)
Comment (rohan47) [reply pkg/operator/ceph/config/network.go:53]: Changing the comment

File: pkg/operator/k8sutil/network_test.go [hunk 11]
```diff
@@ -53,12 +73,15 @@ func TestNetwork_ApplyMultusJSON(t *testing.T) {
...
 
-	assert.Contains(t, objMeta.Annotations, "k8s.v1.cni.cncf.io/networks")
-	assert.Contains(t, objMeta.Annotations["k8s.v1.cni.cncf.io/networks"], `{"name": "macvlan", "interface": "net1"}`)
-	assert.Contains(t, objMeta.Annotations["k8s.v1.cni.cncf.io/networks"], `{"name": "macvlan", "interface": "net2"}`)
+	want := map[string]string{"k8s.v1.cni.cncf.io/networks": `[{"name": "macvlan", "interface": "net1"}, {"name": "macvlan", "interface": "net2"}]`}
+
+	assert.Exactly(t, objMeta.Annotations, want)
```
Comment (rohan47) [inline pkg/operator/k8sutil/network_test.go:84]: @BlaineEXE will have to use `assert.Contains()` as splitting the strings will get complex.
The unit test fails if the order of elements change in the map value string.

**Review by travisn — APPROVED** [review]

File: pkg/operator/k8sutil/network.go [hunk 12]
```diff
@@ -104,6 +109,11 @@ func ApplyMultus(net rookv1.NetworkSpec, objectMeta *metav1.ObjectMeta) error {
 	return nil
 }
 
+// getClusterNwApps returns
```
Comment (leseb) [inline pkg/operator/k8sutil/network.go:112]:
```suggestion
// getClusterNetworkApps returns
```

The description is incomplete.

**Review by leseb — CHANGES_REQUESTED** [review]
One final nit

File: pkg/operator/k8sutil/network.go (hunk 12 above)
Comment (rohan47) [reply pkg/operator/k8sutil/network.go]: Done

**Review by leseb — APPROVED** [review]
